src/lib/runtime/enter/chroot/chroot.c

```diff
@@ -35,19 +36,43 @@
 #include "util/util.h"
 #include "util/message.h"
 #include "util/privilege.h"
+#include "util/registry.h"
 #include "util/config_parser.h"
 
 #include "../../runtime.h"
 
+int pivot_root(const char *new_root, const char *put_old) {
+    return syscall(__NR_pivot_root, new_root, put_old);
+}
 
 int _singularity_runtime_enter_chroot(void) {
     char *container_dir = CONTAINER_FINALDIR;
 
-    singularity_priv_escalate();
     singularity_message(VERBOSE, "Entering container file system root: %s\n", container_dir);
-    if ( chroot(container_dir) < 0 ) { // Flawfinder: ignore (yep, yep, yep... we know!)
-        singularity_message(ERROR, "failed chroot to container at: %s\n", container_dir);
-        ABORT(255);
+
+    singularity_priv_escalate();
+    if ( singularity_registry_get("DAEMON_JOIN") == NULL ) {
+        if ( chdir(container_dir) < 0 ) {
+            singularity_message(ERROR, "Could not chdir to file system root %s: %s\n", container_dir, strerror(errno));
+            ABORT(1);
+        }
+        if ( pivot_root(".", "bin") < 0 ) {
```

Comment:
chroot only changes process current root to container path but doesn't set the real root filesystem in namespace, so when joining instance namespace you access host root file system rather than root container (chroot escape is possible). pivot_root ensure filesystem isolation and make chroot escape impossible

Reply:
Thank you @bauerm97 for your insightful review